### PR TITLE
Qubit tracking

### PR DESCRIPTION
**Pull request by alejandrovgonzalez**

### CODE REVIEW

**Comment by alejandrovgonzalez** [comment]
TODO:
* Check coherence between topology components
* Create Watcher
* Consistent method names between commits
* Consistent and correct documentation

File: src/qccdParser.jl [hunk 1]
```diff
@@ -1,11 +1,31 @@
...
-# Creates a topology using a graph from JSON
+#=  
+    Input -> Json path
+    Output -> Topology
+    Creates a topology using a graph from JSON =#
 function createTopology(path::String)::SimpleDiGraph{Int64}
+    topology::TopologyJSON  = try 
+        _readJSON(path::String)
+    catch err
+        throw(err)
```
Comment (alejandrovgonzalez) [inline src/qccdParser.jl:14]: Catching the error and re-throwing it may not be necessary.

File: src/qccdParser.jl [hunk 2]
```diff
@@ -27,3 +54,75 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
     end
     return graphTopology
 end
+
+#=  
+ =#
+function _createJunctions(shuttles::Array{ShuttleInfoJSON}, junctions::Array{JunctionInfoJSON})::Dict{Int64,Junction}
```
Comment (Hopery) [inline src/qccdParser.jl:60]: The lines are too big. I think it is better if we split them.

File: src/qccdParser.jl [hunk 3]
```diff
@@ -27,3 +54,75 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
...
+#=    Input -> TrapJSON
+    Output -> Dict of qubits
+    Creates a dictionary of qubits using a object TrapJSON, throws an error
+    if qubits are in more than one place at same time. =#
+function _initalizateQubits(trapJSON::TrapJSON)::Dict{String,Qubit}
+    qubits = Dict{String,Qubit}()
+    err = (trapId, qubitPos) -> ArgumentError("Qubit cannot be in two places 
+            at same time when inicializating: " * trapId * ", " * qubitPos * ".")
+
+    for trap in trapJSON.traps
```
Comment (alejandrovgonzalez) [inline src/qccdParser.jl:89]: \in

File: src/qccdParser.jl [hunk 4]
```diff
@@ -27,3 +54,75 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
...
 end
+
+#=  
+ =#
+function _createJunctions(shuttles::Array{ShuttleInfoJSON}, junctions::Array{JunctionInfoJSON})::Dict{Int64,Junction}
+    res = Dict{Int64,Junction}()
+    for j ∈ junctions
+        if haskey(res, j.id)
+            throw(ArgumentError(string("Repeated junction ID:", j.id, ". Make sure the IDs are unique.")))
+        end
```
Comment (Hopery) [inline src/qccdParser.jl:65]: I think is more elegant something like:
```Julia
!haskey(res, j.id) || throw(ArgumentError("Repeated junction ID: " * j.id * " . Make sure the IDs are unique."))
```
Also, message errors must be consistent

File: src/qccdParser.jl [hunk 5]
```diff
@@ -27,3 +54,75 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
...
+        connectedShuttles = Iterators.filter(x -> x.from == j.id || x.to == j.id, shuttles)
+        junctionEnds = Dict(s.id => JunctionEnd() for s ∈ connectedShuttles)
+        try
+            res[j.id] = Junction(j.id, eval(Meta.parse(j.type)), junctionEnds)
+        catch e
+            if e isa UndefVarError
+                throw(ArgumentError(string("Junction type ", j.type, " not supported")))
+            else
+                rethrow(e)
+            end
```
Comment (Hopery) [inline src/qccdParser.jl:75]: Since it's a simple if-else, maybe it can be better to use a ternary.

File: src/types/device.jl [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    ends: Dict with key being the shuttle ID the junction is connected to and value a JunctionEnd
+    @Throws ArgumentError if junction type doesn't match with number of ends. =#
+struct Junction
+    id::Int64 
+    type::JunctionType
+    ends::Dict{String,JunctionEnd}
+    function Junction(id::Int64, type::JunctionType, ends::Dict{String,JunctionEnd})
+        if length(ends) != typesSizes[type]
+            throw(ArgumentError(string("Junction with ID ", id, " of type ",type, " has ",length(ends),
+            " ends. It should have ", typesSizes[type], " ends.")))
```
Comment (Hopery) [inline src/types/device.jl:43]: I think is better to use ` * ` to concatenate strings. Also, maybe the lines are too long.

**Review by Hopery — CHANGES_REQUESTED** [review]
Some changes.

**Review by Hopery — APPROVED** [review]

File: src/qccdParser.jl [hunk 7]
```diff
@@ -15,15 +30,133 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
...
+        catch e
+            e isa UndefVarError ?
+                throw(ArgumentError("Junction type $(j.type) not supported")) : rethrow(e)
+        end
+    end
+    return res
+end
+
+"""
+Creates a dictionary of qubits using a object TrapJSON.
```
Comment (alejandrovgonzalez) [inline src/qccdParser.jl:75]: dead code

File: src/qccdParser.jl [hunk 8]
```diff
@@ -15,15 +30,133 @@ function createTopology(path::String)::SimpleDiGraph{Int64}
...
+              traps[tr.id] = Trap(tr.id,trapJSON.capacity,tr.chain, 
+              TrapEnd(tr.end0.qubit, tr.end0.shuttle), 
+              TrapEnd(tr.end1.qubit, tr.end1.shuttle)), 
+              trapJSON.traps)
+    return traps
+end
+
+"""
+Create the QCCDevStat using dicts.
+Throws error when:
```
Comment (alejandrovgonzalez) [inline src/qccdParser.jl:124]: Error descriptions are wrong, here and in called functions.

**Pull request merged**